runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/DeadlockDetector.java

```diff
@@ -304,23 +307,30 @@ void lockAcquired(Thread owner, ISchedulingRule lock) {
 		 * or conflict with a lock the given lock will acquire implicitly
 		 * (locks are acquired implicitly when a conflicting lock is acquired)
 		 */
-		ArrayList<ISchedulingRule> conflicting = new ArrayList<>(1);
-		//only need two passes through all the locks to pick up all conflicting rules
-		int NUM_PASSES = 2;
-		conflicting.add(lock);
 		graph[threadIndex][lockIndex]++;
-		for (int i = 0; i < NUM_PASSES; i++) {
-			for (int k = 0; k < conflicting.size(); k++) {
-				ISchedulingRule current = conflicting.get(k);
-				for (int j = 0; j < locks.size(); j++) {
-					ISchedulingRule possible = locks.get(j);
-					if (!conflicting.contains(possible) && current.isConflicting(possible)) {
-						conflicting.add(possible);
-						graph[threadIndex][j]++;
-					}
+
+		// first pass tests against lock:
+		Collection<ISchedulingRule> conflicting = Set.of(lock);
+		conflicting = computeConflicting(threadIndex, conflicting);
+
+		// second pass tests also transitive:
+		if (!conflicting.isEmpty()) {
+			conflicting = computeConflicting(threadIndex, conflicting);
+		}
+	}
+
+	private Collection<ISchedulingRule> computeConflicting(int threadIndex, Collection<ISchedulingRule> candidates) {
+		Collection<ISchedulingRule> conflicting = new HashSet<>(candidates);
```

Comment:
You can lazy create the copy on the first found conflict and inti it here with null.

```suggestion
		Collection<ISchedulingRule> conflicting = null;
```

Reply:
to get a NPE on conflicting.contains? Would need to add even more code... premature optimization.